drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,572 @@ impl ServerImpl {
         // trust the message, even though it nominally deserialized correctly.
         // Don't reply.
-        if msg.header.version < version::MIN {
-            ringbuf_entry!(Trace::WrongVersion(msg.header.version));
+        if msg.version() < version::inner::MIN {
+            ringbuf_entry!(Trace::WrongVersion(msg.version()));
             return None;
         }
 
-        let (reply, data_len) = match msg.body {
+        // If the message is something we don't expect or something we cannot
+        // reply to
+        let reserved_framing = Header::MAX_SIZE + Message::MAX_SIZE;
+        let (body, data_len) = match msg.body {
             // These messages should never be sent to us, and we reply
-            // with a `ProtocolError` below.
+            // with a `WrongMessage` below.
             MessageBody::SpRequest(..) => {
                 ringbuf_entry!(Trace::GotSpRequest);
-                (SpResponse::Error(Error::ProtocolError), 0)
+                (MessageBody::Error(ProtocolError::WrongMessage), 0)
             }
             MessageBody::SpResponse(..) => {
                 ringbuf_entry!(Trace::GotSpResponse);
-                (SpResponse::Error(Error::ProtocolError), 0)
+                (MessageBody::Error(ProtocolError::WrongMessage), 0)
             }
             // Nothing implemented yet
             MessageBody::HostResponse(r) => {
                 ringbuf_entry!(Trace::UnexpectedHostResponse(r));
                 return None;
             }
             // Happy path: the host is asking something of us!
-            MessageBody::HostRequest(h) => {
-                match self.handle_host_request(
-                    h,
-                    msg.modules,
-                    data,
-                    &mut tx_data_buf[Message::MAX_SIZE..],
-                ) {
-                    Ok(r) => r,
-                    Err(e) => (SpResponse::Error(e), 0),
-                }
+            MessageBody::HostRequest(h) => self.handle_host_request(
+                h,
+                data,
+                &mut tx_data_buf[reserved_framing..],
+            ),
+            // Nothing implemented yet
+            MessageBody::Error(e) => {
+                ringbuf_entry!(Trace::GotError(e));
+                return None;
             }
         };
 
-        // Serialize the Message into the front of the tx buffer.
-        // Note that we send out _our_ protocol version number, which may differ
-        // from the host's, but should be compatible with it.
-        let header = Header {
-            version: version::CURRENT,
-            ..msg.header
-        };
-        let out = Message {
-            header,
-            modules: msg.modules,
-            body: MessageBody::SpResponse(reply),
-        };
-        let msg_len = hubpack::serialize(tx_data_buf, &out).unwrap();
+        // Serialize the Header into the front of the tx buffer, followed by
```

Comment:
I might make a note that any trailing data is already in the buffer by this point, as part of `handle_host_request()`.